src/compositions/useRoute.ts

```diff
@@ -1,13 +1,46 @@
-import { ExtractRouterRouteParamTypes, RouterRoute } from '@/types'
-import { Identity } from '@/types/utilities'
-
-type Route<T extends RouterRoute> = {
-  name: string,
-  params: ExtractRouterRouteParamTypes<T>,
-  query: unknown,
-  hash: string,
+import { DeepReadonly, watch } from 'vue'
+import { useRouter } from '@/compositions/useRouter'
+import { UseRouteInvalidError } from '@/errors'
+import { RegisteredRouteMap, ResolvedRoute } from '@/types'
+import { combineName } from '@/utilities/combineName'
+
+export function useRoute<TRouteName extends keyof RegisteredRouteMap>(routeName: TRouteName): DeepReadonly<ResolvedRoute<TRouteName>>
+export function useRoute(): DeepReadonly<ResolvedRoute>
+export function useRoute(routeName?: string): DeepReadonly<ResolvedRoute> {
+  const router = useRouter()
+
+  function checkRouteNameIsValid(): void {
+    if (!routeName) {
+      return
+    }
+
+    const routeNames = router.route.matches.map(route => route.name)
+    const routeAncestryNames = getRouteDotNotationNames(routeNames)
+    const routeNameIsValid = routeAncestryNames.includes(routeName)
+
+    if (!routeNameIsValid) {
+      throw new UseRouteInvalidError(routeName, router.route.name)
+    }
+  }
+
+  watch(router.route, () => checkRouteNameIsValid, { immediate: true, deep: true })
```

Comment:
```suggestion
  watch(router.route, checkRouteNameIsValid, { immediate: true, deep: true })
```

Reply:
I'm not sure but we might want to call `checkRouteNameIsValid` manually first rather than using immediate true. Otherwise I think setup will finish. We'll have to test that

Reply:
I'll add a note for myself to explicitly test this when we start testing the router in a test app